vpr/src/device/rr_graph/rr_graph_obj.h

```diff
@@ -0,0 +1,354 @@
+#    include <limits>
+#    include <vector>
+
+/* EXTERNAL library header files go second*/
+#    include "vtr_vector.h"
+#    include "vtr_range.h"
+#    include "vtr_geometry.h"
+#    include "arch_types.h"
+
+/* VPR header files go second*/
+#    include "vpr_types.h"
+#    include "rr_graph_fwd.h"
+
+class RRGraph {
+  public: //Types
+    typedef vtr::vector<RRNodeId, RRNodeId>::const_iterator node_iterator;
+    typedef vtr::vector<RREdgeId, RREdgeId>::const_iterator edge_iterator;
+    typedef vtr::vector<RRSwitchId, RRSwitchId>::const_iterator switch_iterator;
+    typedef vtr::vector<RRSegmentId, RRSegmentId>::const_iterator segment_iterator;
+
+    typedef vtr::Range<node_iterator> node_range;
+    typedef vtr::Range<edge_iterator> edge_range;
+    typedef vtr::Range<switch_iterator> switch_range;
+    typedef vtr::Range<segment_iterator> segment_range;
+
+  public: //Accessors
+    //Aggregates
+    node_range nodes() const;
+    edge_range edges() const;
+    switch_range switches() const;
+    segment_range segments() const;
+
+    //Node attributes
+    size_t node_index(RRNodeId node) const;
+    t_rr_type node_type(RRNodeId node) const;
+    const char* node_type_string(RRNodeId node) const;
+
+    short node_xlow(RRNodeId node) const;
+    short node_ylow(RRNodeId node) const;
+    short node_xhigh(RRNodeId node) const;
+    short node_yhigh(RRNodeId node) const;
+    short node_length(RRNodeId node) const;
+    vtr::Rect<short> node_bounding_box(RRNodeId node) const;
+    /* Node starting and ending points */
+    vtr::Point<size_t> node_start_coordinator(RRNodeId node) const;
+    vtr::Point<size_t> node_end_coordinator(RRNodeId node) const;
+
+    short node_capacity(RRNodeId node) const;
+    short node_fan_in(RRNodeId node) const;
+    short node_fan_out(RRNodeId node) const;
+
+    short node_ptc_num(RRNodeId node) const;
+    short node_pin_num(RRNodeId node) const;
+    short node_track_num(RRNodeId node) const;
+    short node_class_num(RRNodeId node) const;
+
+    short node_cost_index(RRNodeId node) const;
+    e_direction node_direction(RRNodeId node) const;
+    const char* node_direction_string(RRNodeId node) const;
+    e_side node_side(RRNodeId node) const;
+    const char* node_side_string(RRNodeId node) const;
+    float node_R(RRNodeId node) const;
+    float node_C(RRNodeId node) const;
+    RRSegmentId node_segment(RRNodeId node) const; /* get the segment id of a rr_node */
+
+    short node_num_configurable_in_edges(RRNodeId node) const;      /* get the number of configurable input edges of a node */
+    short node_num_non_configurable_in_edges(RRNodeId node) const;  /* get the number of non-configurable input edges of a node */
+    short node_num_configurable_out_edges(RRNodeId node) const;     /* get the number of configurable output edges of a node */
+    short node_num_non_configurable_out_edges(RRNodeId node) const; /* get the number of non-configurable output edges of a node */
```

Comment:
I reworked the functions to call .size() for the ranges. 
This can ease the use with a direct member function, rather than use .size() outside the class.